rfc/graph-pruning-garbage-collection.md

```diff
@@ -0,0 +1,147 @@
+---
+title: "Graph Pruning / Garbage Collection of unused fields"
+author: Jens Neuse
+---
+
+This RFC describes a new workflow to continuously clean up the GraphQL Schema.
+The goal is to remove unused fields from the schema to keep it clean and more maintainable.
+
+## Problem
+
+The GraphQL Schema is a living document that evolves over time.
+Fields are added, modified, and removed.
+Over time, fields that are no longer used are not removed from the schema.
+This leads to a bloated schema that is hard to understand and maintain.
+
+Cosmo already has metrics that show which fields are used or not,
+but there is no automated way or workflow to remove unused fields from the schema.
+
+Removing unused fields through an automated process has the following benefits:
+1. The schema is easier to read understand
+2. The schema is easier to maintain
+3. Unused Code in resolvers can be removed, reducing complexity and technical debt
+4. The schema is smaller, which can lead to better performance
+
+## Solution
+
+The simplest possible solution is to re-use the existing workflow for schema composition checks to also check for unused fields.
+Schema composition checks are already part of the CI/CD pipeline and are executed whenever a new version of a subgraph is published.
+We can extend this workflow to also check for unused fields.
+
+We have several tools to facilitate this.
+We can have strict rules (fail CI) for unused fields, or we can have warnings.
+Depending on how strict the organization wants to operate,
+they can choose the appropriate level of strictness.
+
+In addition, you should be able to immediately remove unused fields or deprecate them first.
+
+## Implementation
+
+First, we need to define the rules for unused fields.
+We might want to have different rules for different environments,
+so it would make sense to define these rules on the namespace level.
+
+Let's create a new namespace:
+
+```shell
+npx wgc namespace create production --unused-fields-warn=30 --unused-fields-fail=60
```

Comment:
I wonder if this could be configured in the same place where linter rules are configured to keep things related similar.

Reply:
Lint rules are per org. I'm assuming that unused field rules would differ across namespaces (environments).